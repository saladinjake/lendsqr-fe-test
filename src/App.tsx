import React from 'react';
import { Route, Routes } from 'react-router-dom';

function App() {
	return (
		<div>
			<Routes>
				<Route path="/" element={<>Login</>} />
				<Route path="/login" element={<>Login</>} />
				<Route path="/dashboard/*" element={<>Dashboard</>}>
					<Route index element={<>Users</>} />
					<Route index path="users" element={<>Users page</>} />
					<Route path="users/:id/profile" element={<>user details</>} />
				</Route>
				<Route path="*" element={<h1>PAGE NOT FOUND</h1>} />
			</Routes>
		</div>
	);
}

export default App;
