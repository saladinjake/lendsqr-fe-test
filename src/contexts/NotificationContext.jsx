import React, { useState } from "react";
const NotificationContext = React.createContext({
  notification: null,
  notificationMessage: null,
  success: (text)=> {},
  error: (text) => {},
  clear: ()=> {},
});

const STATES = {
  SUCCESS: "success",
  ERROR: "error",
};

const NotificationProvider = (props) => {
  const [notification, setNotification] = useState(null);
  const [notificationMessage, setNotificationText] = useState(null);
  const success = (text) => {
    window.scroll(0, 0);
    setNotificationText(text);
    setNotification(STATES.SUCCESS);
  };
  const error = (text) => {
    window.scroll(0, 0);
    setNotificationText(text);
    setNotification(STATES.ERROR);
  };
  const clear = () => {
    setNotificationText(null);
    setNotification(null);
  };
  return (
    <NotificationContext.Provider
      value={{
        success,
        error,
        clear,
        notification,
        notificationMessage,
      }}
    >
      {props.children}
    </NotificationContext.Provider>
  );
};

export { NotificationProvider };
export default NotificationContext;
