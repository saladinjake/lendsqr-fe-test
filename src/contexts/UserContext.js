import React, { createContext, useReducer } from "react";
import  ProductReducer  from "../redux/reducers/courses.reducer";



export const ProductsContext = createContext();

const OnlineCoursesContextProvider = ({ children }) => {
  const courses = [] //get all products from api
  const [state] = useReducer(ProductReducer, courses);

  return (
    <ProductsContext.Provider value={[...state]}>
      {children}
    </ProductsContext.Provider>
  );
};

export default OnlineCoursesContextProvider;
