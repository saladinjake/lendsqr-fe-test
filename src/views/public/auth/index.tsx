import { Route, Routes } from "react-router-dom";
import LoginLanding from "../../../modules/auth/login/LoginForm";


function AuthenticationRouter() {
  return (
    <Routes>
      <Route index path="/" element={<LoginLanding />} />
      <Route path="/login" element={<LoginLanding />} />
    </Routes>
  );
}

export default AuthenticationRouter;
