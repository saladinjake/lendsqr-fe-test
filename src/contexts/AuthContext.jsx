import React, {
  useReducer,
  createContext,
  useContext,
  useState,
  useEffect,
} from "react";
import FingerprintJS from "@fingerprintjs/fingerprintjs";

import * as authService from "../api/auth.service";

import NotificationContext from "./NotificationContext";
import { connect, useDispatch, useSelector } from "react-redux";
import {
  getAcessTokenFromStore,
  getIsAuthenticatedFromStore,
  getUserRolesFromStore,
} from "../redux/reducers/auth.reducer";

import {
  login as loginSuccessfulDispatcher,
  logOut as cleanUserSession,
} from "../redux/actions/auth.action";

import {
  finishAuth,
  deleteToken,
  isSessionExpired,
  getToken,
  getUserData,
  getUserProfile,
  isTokenStillValid,
} from "../utils/tokenConfig";

import { isAuthenticatedByRoles } from "../utils";

const cachedUser = localStorage && JSON?.parse(localStorage.getItem("user"));
const cachedRole =
  localStorage && JSON?.parse(localStorage.getItem("user_roles"));
const cachedToken = localStorage && localStorage.getItem("token");
const cachedAuthorization = localStorage.user && localStorage.token;

const initialState = {
  isAuthenticated: cachedAuthorization ? true : false,
  user: cachedUser ? cachedUser : null,
  token: cachedToken ? cachedToken : null,
  user_roles: cachedRole ? cachedRole : null,
  isAuth: cachedAuthorization ? true : false, //getToken() ? true : false,
  isLoading: false,
  userProfile: null,
  isLoggedIn: false,
  isLoginPending: false,
  loginError: null,
  loadAuthUser: () => isTokenStillValid(),
};

const AuthContext = createContext({
  isAuthenticated: cachedAuthorization ? true : false,
  user: cachedUser ? cachedUser : null,
  token: cachedToken ? cachedToken : null,
  user_roles: cachedRole ? cachedRole : null,
  isAuth: cachedAuthorization ? true : false, //getToken() ? true : false,

  isTokenStillValid: () => isTokenStillValid(),
  isLoading: false,
  userProfile: null,
  login: (credentials) => authService.loginUser(credentials),
  logout: () => {},
  forgetPassword: (payload) => {
   //todo
  },
  changePassword: (payload) => {
    //todo
  },
  updatePassword: (payload) => {
     //todo
  },
  loadAuthUser: () => {},
  modules: [],
  permissions: [],
});

function authReducer(state, action) {
  switch (action.type) {
    case "LOGIN_START":
      return {
        ...state,
        isLoading: true,
      };
    case "LOGIN_SUCCESS":
      return {
        ...state,
        user: action.user,
        isAuth: action.isAuthenticated,
        isLoading: false,
        userProfile: action.userProfile,
        user_roles: action.user_roles,
        isAuthenticated: action.isAuthenticated,
      };
    case "LOGOUT":
      return {
        ...state,
        user: null,
        isAuth: false,
        isLoading: false,
        userProfile: null,
      };
    case "FORGOT_PASSWORD_SUCCESS":
      return {
        ...state,
      };
    case "FORGOT_PASSWORD_FAILED":
      cleanUserSession();
      return {
        ...state,
      };
    case "CHANGE_PASSWORD_FAILED":
      cleanUserSession();
      return {
        ...state,
      };
    case "CHANGE_PASSWORD_SUCCESS":
      cleanUserSession();
      return {
        ...state,
      };

    default:
      return state;
  }
}

function getPublicIp(url) {
  return fetch(url).then((res) => res.text());
}

async function loadAuthUser(state, dispatch, logout) {
  try {
    const isUserStillAuthenticated = await state.loadAuthUser(); //check token profile
    const ssoUserIdentityProfile = isUserStillAuthenticated?.data?.user;
    const ssoUserRole = isUserStillAuthenticated?.data?.user?.roles;
    const adminOnly =
      Array.isArray(ssoUserRole) &&
      ssoUserRole?.every((role) => role?.name === "User");

    if (!isUserStillAuthenticated) {
      logout();
      await cleanUserSession();
    } else if (ssoUserIdentityProfile && adminOnly) {
      if (!isSessionExpired()) {
        let token = getToken();
        let user = getUserData();
        let userProfile = getUserProfile();
        if (token && user) {
          dispatch({
            type: "LOGIN_SUCCESS",
            user,
            isAuth: true,
            userProfile,
            isAuthenticated: true,
          });
        } else {
          // logout();
          await cleanUserSession();
        }
      } else {
        // logout();
        await cleanUserSession();
      }
    }
  } catch (error) {
    console.log(error);
    // logout();
    await cleanUserSession();
  }
}

function AuthProvider(props) {
  const reduxAwareDispatcher = useDispatch();
  const [state, dispatch] = useReducer(authReducer, initialState);
  const notificationCtx = useContext(NotificationContext);
  const [userIp, setMyIp] = useState(null);
  const [message, setMessage] = useState("");
  const isAuthenticatedFromStore = useSelector((state) =>
    getIsAuthenticatedFromStore(state)
  );
  const getPersistenceUser = useSelector((state) =>
    getUserRolesFromStore(state)
  );
  const getPersistenceRoles = useSelector((state) =>
    getUserRolesFromStore(state)
  );
  const getPersistenceAcessTokenFromStore = useSelector((state) =>
    getAcessTokenFromStore(state)
  );
  const dispatchPersistenceLogin = (userData) =>
    reduxAwareDispatcher(loginSuccessfulDispatcher(userData));
  const dispatchPersistenceLogout = () =>
    reduxAwareDispatcher({ type: "LOGIN_FAIL" });

  const [internalsState, setState] = useState(initialState);
  useEffect(() => {
    loadAuthUser(state, dispatch, logout);
  }, [state]);

  const login = async (payload) => {
    try {
      dispatch({ type: "LOGIN_START" });
      const ipaddr = await getPublicIp(
        "https://www.cloudflare.com/cdn-cgi/trace"
      );
      let ipRegex = /[0-9]{1,3}.[0-9]{1,3}.[0-9]{1,3}.[0-9]{1,3}/;
      let ip = ipaddr.match(ipRegex)[0];
      const MaxTimeForRetrials = 3;
      let count = 0;
      const apiResponseData = await authService.loginUser(payload);

      console.log(apiResponseData, ">>>");
      const decoded = apiResponseData?.data;
      const adminRoles = apiResponseData?.data?.user?.roles;
      const hasAdminRoles = isAuthenticatedByRoles("User", getPersistenceRoles);
      if (hasAdminRoles) {
        const userToken = apiResponseData?.data?.access_token;
        const userProfile = {
          id: apiResponseData?.data?.user?.id,
          employeeNumber: apiResponseData?.data?.user?.employeeNumber,
          username: apiResponseData?.data?.user?.username,
          first_name: apiResponseData?.data?.user?.first_name,
          middle_name: apiResponseData?.data?.user?.middle_name,
          last_name: apiResponseData?.data?.user?.last_name,
          email: apiResponseData?.data?.user?.email,
          phone_number: apiResponseData?.data?.user?.phone_number,
          image_url: apiResponseData?.data?.user?.image_url,
          instructor_profile: apiResponseData?.data?.user?.instructor_profile,
          roles: apiResponseData?.data?.user?.roles,
        };

        const tokenExpiry = ""; //decoded?.expires_at || null;
        const fpPromise = await FingerprintJS.load();
        const fpPromiseGetResult = await fpPromise.get();
        const visitorId = fpPromiseGetResult.visitorId;

        const data = {
          userToken,
          auditLogData: {
            SYSTEMIDENTIFIER: "",
            SYSTEMIPADDRESS: ip,
            SYSTEMMACADDRESS: visitorId,
            SYSTEMNAME: "",
            SUBJECTIDENTIFIER: "",
          },
        };

        try {
          const user = apiResponseData.data?.user;
          const ApiResponseStatus =
            apiResponseData?.data?.success &&
            apiResponseData?.data?.status_code;

          if (ApiResponseStatus) {
            finishAuth(userToken, user, visitorId, tokenExpiry, userProfile);
            dispatchPersistenceLogin({
              user,
              isAuthenticated: true,
              access_token: userToken,
              user_roles: getPersistenceRoles,
            });
            notificationCtx.success("Login successful!");
            dispatch({
              type: "LOGIN_SUCCESS",
              user,
              isAuth: isAuthenticatedFromStore, //true,
              userProfile,
              user_roles: getPersistenceRoles,
            });
          } else {
            notificationCtx.error("Invalid credentials. Please Try again");
          }
        } catch (error) {
          count = count + 1;
          dispatch({ type: "LOGOUT" });
          dispatchPersistenceLogout();
          await cleanUserSession();
          notificationCtx.error("Invalid credentials. Please Try again");
        }
        return apiResponseData.data;
      }
      return apiResponseData.data;
    } catch (error) {
      console.log(error);
      notificationCtx.error("Invalid credentials.");
    }
  };

  async function logout() {
    deleteToken();
    dispatch({ type: "LOGOUT" });
    dispatchPersistenceLogout();
    await cleanUserSession();
  }

  async function forgetPassword(payload) {
  
  }

  async function changePassword(payload) {
    
  }

  async function updatePassword(payload) {
    
  }

  return (
    <AuthContext.Provider
      value={{
        user: state.user,
        login,
        logout,
        forgetPassword,
        changePassword,
        updatePassword,
        isAuth: state.isAuth,
        isLoading: state.isLoading,
        loadAuthUser: () => loadAuthUser(state, dispatch, logout),
        userProfile: state.userProfile,
      }}
      {...props}
    >
      {props.children}
    </AuthContext.Provider>
  );
}

const useAuth = () => useContext(AuthContext);
export { AuthContext, AuthProvider, useAuth };
