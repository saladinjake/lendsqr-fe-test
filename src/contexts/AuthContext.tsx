import React, {
  useReducer,
  createContext,
  useContext,
  useState,
  useEffect,
} from "react";
import FingerprintJS from "@fingerprintjs/fingerprintjs";

import * as authService from "../api/auth.service";

import NotificationContext from "./NotificationContext";
import { connect, useDispatch, useSelector } from "react-redux";
import {
  getAcessTokenFromStore,
  getIsAuthenticatedFromStore,
  getUserRolesFromStore,
} from "../redux/reducers/auth.reducer";

import {
  login as loginSuccessfulDispatcher,
  logOut as cleanUserSession,
} from "../redux/actions/auth.action";

import {
  finishAuth,
  deleteToken,
  isSessionExpired,
  getToken,
  getUserData,
  getUserProfile,
  isTokenStillValid,
} from "../utils/tokenConfig";

import { isAuthenticatedByRoles } from "../utils";
import { RootState } from "../redux/reducers";
const cachedUser = localStorage && JSON?.parse(""+localStorage.getItem("user"));
const cachedRole =
  localStorage && JSON?.parse(""+ localStorage.getItem("user_roles"));
const cachedToken = localStorage && localStorage.getItem("token");
const cachedAuthorization = localStorage.user && localStorage.token;

const initialState = {
  isAuthenticated: cachedAuthorization ? true : false,
  user: cachedUser ? cachedUser : null,
  token: cachedToken ? cachedToken : null,
  user_roles: cachedRole ? cachedRole : null,
  isAuth: cachedAuthorization ? true : false, //getToken() ? true : false,
  isLoading: false,
  userProfile: null,
  isLoggedIn: false,
  isLoginPending: false,
  loginError: null,
  loadAuthUser: () => isTokenStillValid(),
};

interface IDecoded {
  email: string;
  family_name: string;
  given_name: string;
  name: string;
  picture: string;
  exp: string;
  designation?: string;
  employeeNumber?: string;
  firstName?: string;
  lastName?: string;
  userRoleID?: string;
  id?: string;
  userName?: string;
  password?: string;
  isSuccessful?: boolean;
}

const AuthContext = createContext({
  isAuthenticated: cachedAuthorization ? true : false,
  user: cachedUser ? cachedUser : null,
  token: cachedToken ? cachedToken : null,
  user_roles: cachedRole ? cachedRole : null,
  isAuth: cachedAuthorization ? true : false, //getToken() ? true : false,

  isTokenStillValid: () => isTokenStillValid(),
  isLoading: false,
  userProfile: null,
  login: (credentials: any) => authService.loginUser(credentials),
  logout: () => {},
  forgetPassword: (payload: any): any =>  {
    
  },
  changePassword: (payload: any): any =>  {
   
  },
  updatePassword: (payload: any): any =>{
   
  },
  loadAuthUser: () => {},
  modules: [],
  permissions: [],
});

function authReducer(state:any, action:any) {
  switch (action.type) {
    case "LOGIN_START":
      return {
        ...state,
        isLoading: true,
      };
    case "LOGIN_SUCCESS":
      return {
        ...state,
        user: action.user,
        isAuth: action.isAuthenticated,
        isLoading: false,
        userProfile: action.userProfile,
        user_roles: action.user_roles,
        isAuthenticated: action.isAuthenticated,
      };
    case "LOGOUT":
      return {
        ...state,
        user: null,
        isAuth: false,
        isLoading: false,
        userProfile: null,
      };
    case "FORGOT_PASSWORD_SUCCESS":
      return {
        ...state,
      };
    case "FORGOT_PASSWORD_FAILED":
      // logOut();
      return {
        ...state,
      };
    case "CHANGE_PASSWORD_FAILED":
      // logOut();
      return {
        ...state,
      };
    case "CHANGE_PASSWORD_SUCCESS":
      // logOut();
      return {
        ...state,
      };

    default:
      return state;
  }
}



async function loadAuthUser(state: any, dispatch: any, logout: any) {

  try {
    const isUserStillAuthenticated = await state.loadAuthUser(); //check token profile
    const ssoUserIdentityProfile = isUserStillAuthenticated?.data?.user;
    const ssoUserRole = isUserStillAuthenticated?.data?.user?.roles;
    const adminOnly = Array.isArray(ssoUserRole) && ssoUserRole?.every((role) => role?.name === "SuperAdmin");

    if (!isUserStillAuthenticated) {
      // logOut();
      await cleanUserSession();
    } else if (ssoUserIdentityProfile && adminOnly) {
      if (!isSessionExpired()) {
        let token = getToken();
        let user = getUserData();
        let userProfile = getUserProfile();
        if (token && user) {
          dispatch({
            type: "LOGIN_SUCCESS",
            user,
            isAuth: true,
            userProfile,
            isAuthenticated: true,
          });
        } else {
          // logout();
          await cleanUserSession();
        }
      } else {
        // logout();
        await cleanUserSession();
      }
    }
  } catch (error) {
    console.log(error);
    // logout();
    await cleanUserSession();
  }
}

function AuthProvider(props: any) {
  const reduxAwareDispatcher = useDispatch();
  const [state, dispatch] = useReducer(authReducer, initialState);
  const notificationCtx = useContext(NotificationContext);
  const [userIp, setMyIp] = useState(null);
  const [message, setMessage] = useState("");
  const isAuthenticatedFromStore = useSelector((state: RootState) =>
    getIsAuthenticatedFromStore(state)
  );
  const getPersistenceUser = useSelector((state: RootState) =>
    getUserRolesFromStore(state)
  );
  const getPersistenceRoles = useSelector((state: RootState) =>
    getUserRolesFromStore(state)
  );
  const getPersistenceAcessTokenFromStore = useSelector((state: RootState) =>
    getAcessTokenFromStore(state)
  );
  const dispatchPersistenceLogin = (userData: any) =>
    reduxAwareDispatcher(loginSuccessfulDispatcher(userData) as any);
  const dispatchPersistenceLogout = () =>
    reduxAwareDispatcher({ type: "LOGIN_FAIL" });

  const [internalsState, setState] = useState(initialState);
  useEffect(() => {
    loadAuthUser(state, dispatch, logout);
  }, [state]);

  const login = async (payload: any) => {
    try {
      dispatch({ type: "LOGIN_START" });
    
      const MaxTimeForRetrials = 3;
      let count = 0;
      const apiResponseData : any= await authService.loginUser(payload);
      const decoded = apiResponseData?.data;
      const adminRoles = apiResponseData?.data?.user?.roles;
      const hasAdminRoles = isAuthenticatedByRoles(
        "SuperAdmin",
        getPersistenceRoles
      );
      if (hasAdminRoles) {
        const userToken = apiResponseData?.data?.access_token;
        const userProfile = {
          id: apiResponseData?.data?.user?.id,
          employeeNumber: apiResponseData?.data?.user?.employeeNumber,
          username: apiResponseData?.data?.user?.username,
          first_name: apiResponseData?.data?.user?.first_name,
          middle_name: apiResponseData?.data?.user?.middle_name,
          last_name: apiResponseData?.data?.user?.last_name,
          email: apiResponseData?.data?.user?.email,
          phone_number: apiResponseData?.data?.user?.phone_number,
          image_url: apiResponseData?.data?.user?.image_url,
          instructor_profile: apiResponseData?.data?.user?.instructor_profile,
          roles: apiResponseData?.data?.user?.roles,
        };
     
        const tokenExpiry = ""; //decoded?.expires_at || null;
        const fpPromise = await FingerprintJS.load();
        const fpPromiseGetResult = await fpPromise.get();
        const visitorId = fpPromiseGetResult.visitorId;

        const data = {
          userToken,
          auditLogData: {
            SYSTEMIDENTIFIER: "",
            SYSTEMIPADDRESS: "192.168.1.1",
            SYSTEMMACADDRESS: visitorId,
            SYSTEMNAME: "",
            SUBJECTIDENTIFIER: "",
          },
        };

        try {
          const user = apiResponseData.data?.user;
          const ApiResponseStatus =
            apiResponseData?.data?.success &&
            apiResponseData?.data?.status_code;

          if (ApiResponseStatus) {
            finishAuth(userToken, user, visitorId, tokenExpiry, userProfile);
            dispatchPersistenceLogin({
              user,
              isAuthenticated: true,
              access_token: userToken,
              user_roles: getPersistenceRoles,
            });
            notificationCtx.success("Login successful!");
            dispatch({
              type: "LOGIN_SUCCESS",
              user,
              isAuth: isAuthenticatedFromStore, //true,
              userProfile,
              user_roles: getPersistenceRoles,
            });
          } else {
            notificationCtx.error("Invalid credentials. Please Try again");
          }
        } catch (error) {
          count = count + 1;
          dispatch({ type: "LOGOUT" });
          dispatchPersistenceLogout();
          await cleanUserSession();
          notificationCtx.error("Invalid credentials. Please Try again");
        }
        return apiResponseData.data;
      }
      return apiResponseData.data;
    } catch (error) {
      console.log(error);
      notificationCtx.error("Invalid credentials.");
    }
  };

  async function logout() {
    deleteToken();
    dispatch({ type: "LOGOUT" });
    dispatchPersistenceLogout();
    await cleanUserSession();
  }

  async function forgetPassword(payload:any) {
   
  }

  async function changePassword(payload: any) {
   
  }

  async function updatePassword(payload: any) {
    
  }

  return (
    <AuthContext.Provider
      value={{
        user: state.user,
        login,
        logout,
        forgetPassword,
        changePassword,
        updatePassword,
        isAuth: state.isAuth,
        isLoading: state.isLoading,
        loadAuthUser: () => loadAuthUser(state, dispatch, logout),
        userProfile: state.userProfile,
      }}
      {...props}
    >
      {props.children}
    </AuthContext.Provider>
  );
}

const useAuth = () => useContext(AuthContext);
export { AuthContext, AuthProvider, useAuth };