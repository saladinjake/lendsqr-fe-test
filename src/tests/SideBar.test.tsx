import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";
import { rest } from "msw";
import { setupServer } from "msw/node";
import App from "../routes/index";

import React from "react";

import { BrowserRouter } from "react-router-dom";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import { RouterProvider } from "react-router-dom";
import { enableFetchMocks } from "jest-fetch-mock";

import { Provider as Provisioner } from "react-redux";
import Store from "../redux/store";
import { NotificationProvider } from "../contexts/NotificationContext";
import NotificationBar from "../components/shared/Notifier";
//import { ReactQueryDevtools } from "@tanstack/react-query-devtools";
import AppNavigator from "../routes";
import { AuthProvider } from "../contexts/AuthContext";
import { loginUser, logoutUser } from "../api/auth.service";
import { getAuthProfile } from "../api/user.service";




import "@testing-library/jest-dom";

//components
import SideBar from "../components/shared/Sidebar/index";;

const modeUrl =
  process.env.NODE_ENV == "development"
    ? "http://localhost:3000"
    : "https://victorjuwa-lendsqr-fe-test.vercel.app";

const mockServer = setupServer(
  rest.get(`${process.env.REACT_APP_API_URL}/users`, (req, res, ctx) => {
    return res(
      ctx.json({
        firstName: "victor",
        lastName: "Juwa",
        role: "Lead developer",
      })
    );
  })
);

beforeAll(() => mockServer.listen());
afterEach(() => mockServer.resetHandlers());
afterAll(() => mockServer.close());


describe("Nav Bar Component", () => {
    const initialState = { output: 10 };
  const mockStore = configureStore();
  let store;
  it("renders without crashing on large screen", () => {

    store = mockStore(initialState);
    const component = render(
      <BrowserRouter>
        <SideBar isOpen={false} />
      </BrowserRouter>
    );
    expect(component.container).toMatchSnapshot();

  
  });

  it('should display Side Component On the browser', async () => {
    render(	<BrowserRouter>
              <SideBar isOpen={false}  />
        </BrowserRouter>
    );
    waitFor(() => {
        const nav = screen.getByTestId('nav-responsive');
        expect(nav).toBeInTheDocument();
    });
});

  it('should contain side bar items', async () => {
  	const component =render(	<BrowserRouter>
  		    <SideBar isOpen={false}  />
  		</BrowserRouter>
  	);
  	waitFor(() => {
        const linkField = screen.getByText(/Customers/i);  
        const version = screen.getByText(/v1.2.0/i);
        expect(linkField).toBeInTheDocument();
        expect(version).toBeInTheDocument();
        let linkList: any = component.container.querySelector('.test-links') ;
        linkList.forEach( (item:any, index: number)  => {
           expect(item).toHaveAttribute('class', 'test-links')
         })
  	});
  });
});