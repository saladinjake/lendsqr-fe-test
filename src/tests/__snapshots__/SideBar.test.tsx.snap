// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav Bar Component renders without crashing on large screen 1`] = `
<div>
  <div
    class="side-bar false"
  >
    <div
      class="branding"
    >
      <img
        alt="logo"
        class="logo"
        src="/images/logo.svg"
      />
    </div>
    <div
      class="sitemaps-wrapper"
    >
      <ul
        class="side-bar-navigation-menu"
      >
        <div>
          <a
            class="test-links"
            href="/"
            title="0-link"
          >
            <li
              class="menu-item false"
            >
              <img
                alt="Switch Organization"
                src="/images/icons/nav-icon-4.svg"
              />
              <span>
                Switch Organization
              </span>
              <span>
                <img
                  alt=""
                  src="/images/icons/down-arrow.svg"
                />
              </span>
            </li>
          </a>
        </div>
        <div>
          <a
            class="test-links"
            href="/dashboard"
            title="1-link"
          >
            <li
              class="menu-item false"
            >
              <img
                alt="Dashboard"
                src="/images/icons/home.svg"
              />
              <span>
                Dashboard
              </span>
            </li>
          </a>
        </div>
        <div>
          <li
            class="link-description"
          >
            <span>
              Customers
            </span>
          </li>
        </div>
        <div>
          <a
            class="test-links"
            href="/dashboard/users"
            title="3-link"
          >
            <li
              class="menu-item false"
            >
              <img
                alt="Users"
                src="/images/icons/nav-icon-4.svg"
              />
              <span>
                Users
              </span>
            </li>
          </a>
        </div>
        <div>
          <a
            class="test-links"
            href="/"
            title="4-link"
          >
            <li
              class="menu-item false"
            >
              <img
                alt="Guarantors"
                src="/images/icons/nav-icon-5.svg"
              />
              <span>
                Guarantors
              </span>
            </li>
          </a>
        </div>
        <div>
          <a
            class="test-links"
            href="/"
            title="5-link"
          >
            <li
              class="menu-item false"
            >
              <img
                alt="Loan"
                src="/images/icons/nav-icon-6.svg"
              />
              <span>
                Loan
              </span>
            </li>
          </a>
        </div>
        <div>
          <a
            class="test-links"
            href="/"
            title="6-link"
          >
            <li
              class="menu-item false"
            >
              <img
                alt="Decision Models"
                src="/images/icons/nav-icon-7.svg"
              />
              <span>
                Decision Models
              </span>
            </li>
          </a>
        </div>
        <div>
          <a
            class="test-links"
            href="/"
            title="7-link"
          >
            <li
              class="menu-item false"
            >
              <img
                alt="Savings"
                src="/images/icons/nav-icon-8.svg"
              />
              <span>
                Savings
              </span>
            </li>
          </a>
        </div>
        <div>
          <a
            class="test-links"
            href="/"
            title="8-link"
          >
            <li
              class="menu-item false"
            >
              <img
                alt="Loan Requests"
                src="/images/icons/nav-icon-9.svg"
              />
              <span>
                Loan Requests
              </span>
            </li>
          </a>
        </div>
        <div>
          <a
            class="test-links"
            href="/"
            title="9-link"
          >
            <li
              class="menu-item false"
            >
              <img
                alt="Whitelist"
                src="/images/icons/nav-icon-10.svg"
              />
              <span>
                Whitelist
              </span>
            </li>
          </a>
        </div>
        <div>
          <a
            class="test-links"
            href="/"
            title="10-link"
          >
            <li
              class="menu-item false"
            >
              <img
                alt="Karma"
                src="/images/icons/nav-icon-11.svg"
              />
              <span>
                Karma
              </span>
            </li>
          </a>
        </div>
        <div>
          <li
            class="link-description"
          >
            <span>
              Businesses
            </span>
          </li>
        </div>
        <div>
          <a
            class="test-links"
            href="/"
            title="12-link"
          >
            <li
              class="menu-item false"
            >
              <img
                alt="Organization"
                src="/images/icons/nav-icon-13.svg"
              />
              <span>
                Organization
              </span>
            </li>
          </a>
        </div>
        <div>
          <a
            class="test-links"
            href="/"
            title="13-link"
          >
            <li
              class="menu-item false"
            >
              <img
                alt="Loan Products"
                src="/images/icons/nav-icon-14.svg"
              />
              <span>
                Loan Products
              </span>
            </li>
          </a>
        </div>
        <div>
          <a
            class="test-links"
            href="/"
            title="14-link"
          >
            <li
              class="menu-item false"
            >
              <img
                alt="Savings Products"
                src="/images/icons/nav-icon-15.svg"
              />
              <span>
                Savings Products
              </span>
            </li>
          </a>
        </div>
        <div>
          <a
            class="test-links"
            href="/"
            title="15-link"
          >
            <li
              class="menu-item false"
            >
              <img
                alt="Fees and Charges"
                src="/images/icons/nav-icon-16.svg"
              />
              <span>
                Fees and Charges
              </span>
            </li>
          </a>
        </div>
        <div>
          <a
            class="test-links"
            href="/"
            title="16-link"
          >
            <li
              class="menu-item false"
            >
              <img
                alt="Transactions"
                src="/images/icons/nav-icon-17.svg"
              />
              <span>
                Transactions
              </span>
            </li>
          </a>
        </div>
        <div>
          <a
            class="test-links"
            href="/"
            title="17-link"
          >
            <li
              class="menu-item false"
            >
              <img
                alt="Services"
                src="/images/icons/nav-icon-18.svg"
              />
              <span>
                Services
              </span>
            </li>
          </a>
        </div>
        <div>
          <a
            class="test-links"
            href="/"
            title="18-link"
          >
            <li
              class="menu-item false"
            >
              <img
                alt="Service Account"
                src="/images/icons/nav-icon-19.svg"
              />
              <span>
                Service Account
              </span>
            </li>
          </a>
        </div>
        <div>
          <a
            class="test-links"
            href="/"
            title="19-link"
          >
            <li
              class="menu-item false"
            >
              <img
                alt="Settlements"
                src="/images/icons/nav-icon-20.svg"
              />
              <span>
                Settlements
              </span>
            </li>
          </a>
        </div>
        <div>
          <a
            class="test-links"
            href="/"
            title="20-link"
          >
            <li
              class="menu-item false"
            >
              <img
                alt="Reports"
                src="/images/icons/nav-icon-21.svg"
              />
              <span>
                Reports
              </span>
            </li>
          </a>
        </div>
        <div>
          <li
            class="link-description"
          >
            <span>
              Settings
            </span>
          </li>
        </div>
        <div>
          <a
            class="test-links"
            href="/"
            title="22-link"
          >
            <li
              class="menu-item false"
            >
              <img
                alt="Preferences"
                src="/images/icons/nav-icon-23.svg"
              />
              <span>
                Preferences
              </span>
            </li>
          </a>
        </div>
        <div>
          <a
            class="test-links"
            href="/"
            title="23-link"
          >
            <li
              class="menu-item false"
            >
              <img
                alt="Fees and Pricing"
                src="/images/icons/nav-icon-24.svg"
              />
              <span>
                Fees and Pricing
              </span>
            </li>
          </a>
        </div>
        <div>
          <a
            class="test-links"
            href="/"
            title="24-link"
          >
            <li
              class="menu-item false"
            >
              <img
                alt="Audit Logs"
                src="/images/icons/nav-icon-25.svg"
              />
              <span>
                Audit Logs
              </span>
            </li>
          </a>
        </div>
        <div>
          <a
            class="test-links"
            href="/"
            title="25-link"
          >
            <li
              class="menu-item false"
            >
              <img
                alt="Systems Messages"
                src="/images/icons/nav-icon-26.svg"
              />
              <span>
                Systems Messages
              </span>
            </li>
          </a>
        </div>
      </ul>
      <div
        class="cta-revoke"
      >
        <div>
          <img
            alt="logout icon"
            src="/images/icons/logout-icon.svg"
          />
          <span>
            Logout
          </span>
        </div>
        <span
          class="info-app"
        >
          v1.2.0
        </span>
      </div>
    </div>
  </div>
</div>
`;
